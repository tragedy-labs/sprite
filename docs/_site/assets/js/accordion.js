/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
