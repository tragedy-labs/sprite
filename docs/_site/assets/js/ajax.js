// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
