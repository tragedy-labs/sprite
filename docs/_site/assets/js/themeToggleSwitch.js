function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
