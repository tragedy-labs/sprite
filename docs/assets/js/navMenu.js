document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
