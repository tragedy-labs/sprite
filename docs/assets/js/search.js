import data from "/assets/js/searchData.js";
import serializedIndex from "./assets/js/serialized.js";
//import lunr from "lunr"; // assuming lunr is a separate module

// Constants
const MIN_SEARCH_LENGTH = 3;

// State definition
const state = {
  filters: {
    class: true,
    method: true,
    type: true,
  },
  results: {
    filtered: [],
    full: [],
  },
  visible: false,
};

// Elements for reference
const elements = {
  input: null,
  resultsBox: null,
  resultList: null,
  numberOfResults: null,
  numberResultsShowing: null,
  backdrop: null,
  filters: null,
  searchLabel: null,
  close: null,
};

// Initialize search
async function initializeSearch() {
  const root = document.getElementById("search_button");
  elements.input = document.getElementById("searchInput");
  elements.resultsBox = document.getElementById("searchResults");
  elements.resultList = document.getElementById("resultList");
  elements.numberOfResults = document.getElementById("numberOfResults");
  elements.numberResultsShowing = document.getElementById("numberResultsShowing");
  elements.filters = document.getElementById("searchFilters")
    .getElementsByTagName("input");
  elements.searchLabel = document.getElementById("searchLabel");
  elements.close = elements.resultsBox.querySelector("#close_search_dialog");

  // Add event listeners
  elements.input.addEventListener("input", search);
  elements.close.addEventListener("click", hideSearchBox);

  const openSearch = document.getElementById('search_button');

  openSearch.addEventListener('click', showResultsBox);
  document.addEventListener('keydown', (e)=>{
    if (e.key === '/') {
      e.preventDefault();
      showResultsBox();
    };
  }, false);

  // Load the index
  const index = lunr.Index.load(serializedIndex);

  // Initialize search filter toggles
  for (let i = 0; i < elements.filters.length; ++i) {
    elements.filters[i].addEventListener("change", () => {
      state.filters[elements.filters[i].value] = elements.filters[i].checked;
      updateResultList();
    });
  }

  // Define functions
  function appendBackdrop() {
    elements.backdrop = document.createElement("div");
    elements.backdrop.id = "backdrop";
    elements.backdrop.addEventListener("click", hideSearchBox);
    document.body.appendChild(elements.backdrop);
  }

  function removeBackdrop() {
    elements.backdrop.removeEventListener("click", appendBackdrop);
    elements.backdrop.remove();
    elements.backdrop = undefined;
  }

  function handleKeyboardSelection() {
    const items = elements.list.querySelectorAll("a");
    if (items.length > 0) {
      if (document.activeElement === elements.input) {
        e.preventDefault();
        items[0].focus();
      } else {
        for (let i = 0; i < results.length; i++) {
          if (items[i] === document.activeElement) {
            e.preventDefault();
            if (i + 1 < items.length) {
              items[i + 1].focus();
            } else {
              elements.input.focus();
            }
            break;
          }
        }
      }
    }
  }

  function checkKeyPress(e) {
    if (e.key === "Tab") {
      handleKeyboardSelection(e);
    }
    if (e.key === "Escape") {
      hideSearchBox();
    }
  }

  function showResultsBox() {
    appendBackdrop();
    state.visible = true;
    elements.resultsBox.classList.remove("hidden");
    elements.input.focus();
    document.addEventListener("keydown", checkKeyPress);
  }

  function hideSearchBox() {
    state.visible = false;
    elements.resultsBox.classList.add("hidden");
    document.removeEventListener("keydown", checkKeyPress);
    removeBackdrop();
  }

  function updateResultList() {
    state.results.filtered = [];
    elements.resultList.innerHTML = "";
    // elements.numberOfResults.innerHTML = state.results.full.length;
    // elements.numberResultsShowing.innerHTML = state.results.filtered.length;

    if (state.results.full.length > 0) {
      state.results.full.forEach((result) => {
        const doc = data.find((doc) => doc.name === result.ref);
        const type = doc.type === "interface" ? "type" : doc.type;
        if (state.filters[type]) {
          state.results.filtered.push(result.ref);
          const section = document.createElement("section");
          const desc = document.createElement("p");
          const link = document.createElement("a");
          const li = document.createElement("li");
          li.classList.add(doc.type);

          link.href = `/sprite/doc/${doc.name}`;
          link.className = "searchResultLink";
          link.addEventListener("click", () => {});
          link.textContent = doc.parent
            ? `${doc.parent}.${doc.name}`
            : doc.name;

          desc.textContent = doc.desc;

          section.appendChild(link);
          section.appendChild(desc);

          li.appendChild(section);

          elements.resultList.appendChild(li);
        }
      });

      //elements.numberResultsShowing.innerHTML = state.results.filtered.length;
    }
  }

  function displayResults() {
    if (!state.visible) {
      showResultsBox();
    }
    updateResultList();
  }

  function search() {
    if (elements.input.value.length > 0) {
      elements.searchLabel.classList.add("hidden");
    } else {
      elements.searchLabel.classList.remove("hidden");
    }

    if (elements.input.value.length >= MIN_SEARCH_LENGTH) {
      const initial = index.search(elements.input.value);

      if (initial.length === 0) {
        const wildcardResults = index.search(`${elements.input.value}*`);
        state.results.full =
          wildcardResults.length > 0
            ? wildcardResults
            : index.search(`*${elements.input.value}*`);
      } else {
        state.results.full = initial;
      }

      displayResults();
    }
  }
}

document.addEventListener("DOMContentLoaded", initializeSearch);