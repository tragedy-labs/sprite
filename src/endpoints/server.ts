export const endpoints = {
  ready: '/api/v1/ready',
  command: '/api/v1/server',
  databases: '/api/v1/databases',
  exists: '/api/v1/exists'
} as const;
