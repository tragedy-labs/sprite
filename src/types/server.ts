export interface ISpriteCreateArcadeUser {
  /** The username of the database user to create */
  username: string;
  /** The password of the database user to create */
  password: string;
  /**
   * An object containing databases to add the user to, and the
   * permisisions to grant them \
   * (i.e. `{ myDatabase: "ADMIN" }`)
   */
  databases: { [key: string]: string };
}

/** Describes an event on the ArcadeDB server */
export type SpriteArcadeServerEvent = {
  time: string;
  type: string;
  message: string;
};

/** The ArcadeDB Server Events Log */
export type SpriteArcadeServerEvents = {
  /** The list of server events */
  events: SpriteArcadeServerEvent[];
  /** Event log files */
  files: string[];
};

/**
 * The level of detail that should be returned for a
 * `SpriteServer.getInformation()` request
 * @default 'default'
 */
export type ArcadeServerInformationLevel = 'basic' | 'default' | 'cluster';

/**
 * The metadata returned from the server on every operation.
 */
export type ArcadeServerBasicInformation = {
  /** The user who performed the operation */
  user: string;
  /** The version of ArcadeDB on the server */
  version: string;
  /** The name of the server */
  serverName: string;
};

export type ArcadeServerInfoSetting = {
  key: string;
  value: unknown;
  description: string;
  overridden: boolean;
  default: unknown;
};

export type ArcadeServerInfoMetrics = {
  meters: object;
  events: { errors: number; warnings: number; info: number; hints: number };
  profiler: {
    readCacheUsed: object;
    writeCacheUsed: object;
    cacheMax: object;
    pagesRead: object;
    pagesWritten: object;
    pagesReadSize: object;
    pagesWrittenSize: object;
    pageFlushQueueLength: object;
    asyncQueueLength: object;
    asyncParallelLevel: object;
    pageCacheHits: object;
    pageCacheMiss: object;
    totalOpenFiles: object;
    maxOpenFiles: object;
    walPagesWritten: object;
    walBytesWritten: object;
    walTotalFiles: 8;
    concurrentModificationExceptions: object;
    txCommits: object;
    txRollbacks: object;
    createRecord: object;
    readRecord: object;
    updateRecord: object;
    deleteRecord: object;
    queries: object;
    commands: object;
    scanType: object;
    scanBucket: object;
    iterateType: object;
    iterateBucket: object;
    countType: object;
    countBucket: object;
    evictionRuns: object;
    pagesEvicted: object;
    readCachePages: object;
    writeCachePages: object;
    indexCompactions: object;
    diskFreeSpace: object;
    diskTotalSpace: object;
    diskFreeSpacePerc: object;
    gcTime: object;
    ramHeapUsed: object;
    ramHeapMax: object;
    ramHeapAvailablePerc: object;
    ramOsUsed: object;
    ramOsTotal: object;
    cpuLoad: object;
    jvmSafePointTime: object;
    jvmSafePointCount: object;
    jvmAvgSafePointTime: object;
    totalDatabases: object;
    cpuCores: object;
    configuration: object;
  };
};

export type ArcadeServerDefaultInformation = ArcadeServerBasicInformation & {
  metrics: ArcadeServerInfoMetrics;
  settings: ArcadeServerInfoSetting[];
};

export type ArcadeServerClusterInformation = object;

export type ArcadeServerInformation<T extends ArcadeServerInformationLevel> = {
  basic: ArcadeServerBasicInformation;
  default: ArcadeServerDefaultInformation;
  cluster: ArcadeServerClusterInformation;
}[T];
