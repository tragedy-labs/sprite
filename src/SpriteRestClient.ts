import { ArcadeDatabaseError } from './errors/ArcadeDatabaseError.js';
import { ArcadeFetchError } from './errors/ArcadeFetchError.js';
import { endpoints } from './endpoints/server.js';
import {
  ISpriteConnection,
  ISpriteRestClientConnectionParameters
} from './types/client.js';

/**
 * Handles the fetch implementation and formatting of credentials.
 */
class SpriteRestClient {
  connection: ISpriteConnection;
  constructor(parameters: ISpriteRestClientConnectionParameters) {
    this.connection = this.getConnection(parameters);
  }
  /**
   * Formats the connection parameters into a configuration object that can be used to interact with the ArcadeDB server.
   * @param params The username, password, and address of the database.
   * @returns The configuration necessary to have REST interactions with the ArcadeDB server.
   * @throws `Error` if valid parameter types are not received, or something else causes an error when building the object.
   */
  getConnection = (
    params: ISpriteRestClientConnectionParameters
  ): ISpriteConnection => {
    try {
      const base64EncodedCredentials = this.encodeCredentials(
        params.username,
        params.password
      );
      return {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Basic ${base64EncodedCredentials}`
        },
        address: params.address
      };
    } catch (error) {
      throw new Error('There was an error building the connection object.', {
        cause: error
      });
    }
  };
  /**
   * Base64 Encodes the `username` and `password` for use in the `Authorization` header.
   * This is a cross-platform solution for nodejs, or browser apps.
   * `btoa()` is marked as deprecated in nodejs, so this solution avoids deprecation
   * warnings.
   * @param username The string username to encode.
   * @param password The string password to encode.
   * @returns The base64 encoded credentials.
   * @example
   * const encoded = await this.encodeCredentials("root", "password");
   * console.log(encoded);
   * // "cm9vdDpwYXNzd29yZA=="
   */
  encodeCredentials = (username: string, password: string): string => {
    // TODO: I have not tried this is Deno or a browser environment.
    try {
      const credentials = `${username}:${password}`;
      if (typeof window !== 'undefined' && typeof window.btoa === 'function') {
        return window.btoa(credentials);
      } else if (typeof Buffer === 'function') {
        return Buffer.from(credentials, 'utf-8').toString('base64');
      } else {
        throw new Error(
          'Could not determine the encoding function to use given the current JavaScript environment.'
        );
      }
    } catch (error) {
      throw new Error('Could not base64 encode the credentials.', {
        cause: error
      });
    }
  };
  /**
   * A convenience method for making REST calls to the ArcadeDB server.\
   * It allows for relative paths to be used as endpoints, and automatically appends
   * the server address and the `Authorization` header to the request.\
   * It also handles the response from the server, throwing an error if the server returns
   * a non-200 status code.
   * @param endpoint The relative path to the endpoint (i.e. /api/v1/ready)
   * @param options The request options for the fetch api.
   * @returns The fetch response
   */
  fetch = async (
    endpoint: string,
    { method = 'GET', body, headers }: RequestInit = {}
  ): Promise<Response> => {
    const response = await fetch(`${this.connection.address}${endpoint}`, {
      method,
      headers: headers
        ? {
            ...this.connection.headers,
            ...headers
          }
        : this.connection.headers,
      body
    });
    switch (response.status) {
      case 403:
      case 404:
        throw new ArcadeFetchError(response);
      default:
        return response;
    }
  };
  /**
   * Fetch API wrapper for making REST calls to the ArcadeDB server and returning a
   * JavaScript object of type `<T>`. This method typically returns the `result` property
   * of the response object from ArcadeDB, but may also throw an `Error` depending on the
   * outcome of the database operation.
   * @param endpoint
   * @param requestParameters
   * @returns The response from the server, typically the `result` property of the response.
   * @example
   * const response = await sprite.arcadeJson<boolean>(endpoint, parameters);
   * console.log(response);
   * // true
   */
  fetchJson = async <T>(
    endpoint: string,
    parameters?: RequestInit
  ): Promise<T> => {
    const response = await this.fetch(endpoint, parameters);
    const json = await response.json();
    if ('result' in json) {
      return json.result;
    } else if ('error' in json) {
      throw new ArcadeDatabaseError(json);
    } else {
      return json;
    }
  };
  /**
   * Issue a command to the server.
   * @param command The command to perform on the server.
   * @returns The response from the server
   * @throws If the command could not be executed
   */
  serverCommand = async <T>(command: string): Promise<T> => {
    try {
      const body = JSON.stringify({
        command
      });
      return await this.fetchJson<T>(endpoints.command, {
        method: 'POST',
        body
      });
    } catch (error) {
      throw new Error(
        `Encountered an error when sending a command to the server.`,
        { cause: error }
      );
    }
  };
}

export { SpriteRestClient };
