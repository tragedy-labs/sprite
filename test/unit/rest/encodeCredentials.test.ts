import { SpriteRestClient } from '../../../src/SpriteRestClient.js';
import { endpoints } from '../../../src/endpoints/server.js';
import { variables } from '../../variables.js';
import { client } from './utilities/testClient.js';

// TODO: These tests are kind of a skeleton for browser / node.
// I'm guessing how to test this in the browser / deno from a
// nodejs environment.

describe('SpriteBase.encodeCredentials()', () => {
  let originalWindow: Window & typeof globalThis;
  let originalBuffer: typeof Buffer;
  beforeEach(() => {
    originalWindow = global.window;
    originalBuffer = global.Buffer;
  });
  afterEach(() => {
    global.window = originalWindow;
    global.Buffer = originalBuffer;  
  });
  const credentialString = `${variables.username}:${variables.password}`;
  it('should properly encode the credentials in a nodejs environment (such as those passed to this.fetch in the `Authorization` header)', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValue(new Response());
    jest.spyOn(Buffer, 'from');

    client.encodeCredentials(variables.username, variables.password);

    expect(Buffer.from).toHaveBeenCalledWith(credentialString, 'utf-8');
  });

  it('should properly encode the credentials in a browser/deno environment (such as those passed to this.fetch in the `Authorization` header)', async () => {
    const originalWindow = global.window;
    global.window = {
      btoa: jest.fn((input) => Buffer.from(input, 'utf-8').toString('base64'))
    } as unknown as Window & typeof globalThis;

    jest.spyOn(global, 'fetch').mockResolvedValue(new Response());

    // We create a new `SpriteBase` instance with the `address`, `username`, and `password`
    // to test that the credentials are being encoded when client is created. Typically
    // tests just use the variable `testClient`, but in this case we need a new instance.
    const client = new SpriteRestClient({
      address: variables.address,
      username: variables.username,
      password: variables.password
    });

    await client.fetch(endpoints.ready);

    // The encodeCredentials method is written so that it uses `btoa`
    // if it's available on the window object (Browser and Deno), and
    // falls back to `Buffer.from` if it's not. We expect `btoa` is
    // called in this test because we've set it on the window object.
    expect(global.window.btoa).toHaveBeenCalledWith(credentialString);

    expect(fetch).toHaveBeenCalledWith(
      `${variables.address}${endpoints.ready}`,
      {
        method: 'GET',
        headers: {
          Authorization: `Basic ${Buffer.from(
            credentialString,
            'utf-8'
          ).toString('base64')}`,
          'Content-Type': 'application/json'
        }
      }
    );

    global.window = originalWindow;
  });

  it('should throw an error if no password is provided', async () => {
    expect(
      () =>
        // @ts-expect-error - We're testing the error thrown when the password is not provided
        new SpriteBase({
          username: variables.username,
          address: variables.address
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw an error if no username is provided', async () => {
    expect(
      () =>
        // @ts-expect-error - We're testing the error thrown when the username is not provided
        new SpriteBase({
          password: variables.password,
          address: variables.address
        })
    ).toThrowErrorMatchingSnapshot();
  });

  test('should throw an error when Buffer is not available', () => {
    global.window = undefined as unknown as typeof global.window;
    global.Buffer = undefined as unknown as typeof Buffer;

    expect(() => client.encodeCredentials('user', 'pass')).toThrowErrorMatchingSnapshot()
  });

  test('should throw an error when Buffer fails', () => {
    global.window = undefined as unknown as typeof global.window;
    global.Buffer = {
      from: jest.fn(() => { throw new Error('Buffer error'); })
    } as unknown as typeof Buffer;

    expect(() => client.encodeCredentials('user', 'pass')).toThrowErrorMatchingSnapshot();
  });



});
