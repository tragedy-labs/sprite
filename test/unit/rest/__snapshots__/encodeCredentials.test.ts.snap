// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteBase.encodeCredentials() should throw an error if no password is provided 1`] = `"SpriteBase is not defined"`;

exports[`SpriteBase.encodeCredentials() should throw an error if no username is provided 1`] = `"SpriteBase is not defined"`;

exports[`SpriteBase.encodeCredentials() should throw an error when Buffer fails 1`] = `"Could not base64 encode the credentials."`;

exports[`SpriteBase.encodeCredentials() should throw an error when Buffer is not available 1`] = `"Could not base64 encode the credentials."`;
